### ISSUE TITLE
Library variables recomputed on each module call

### ISSUE THREAD

**Issue by OskarLinde** [openscad/openscad#782]
Here is a simplified synthetic case to illustrate the problem:

**_Snippet 1**_

```
lookup_table = expensive_computation();
function f(x) = interpolate(x, using = lookup_table);
```

**_Snippet 2**_

```
for (i=[1:1000]) echo(f(x));
```

If these two snippets are placed in the same file, everything executes as expected. The expensive_computation() is performed once and the cheap `interpolate` call does the work as intended. Now to the problem: if we instead put Snippet 1 in its own library: `lib.scad`, and call it like this

```
use<lib.scad>
for (i=[1:1000]) echo(f(x));
```

the expensive_computation() call will suddenly be performed 1000 times – causing a dramatic slowdown.

While this could all be as intended (there is a fringe use-case, see below), I think it is more likely an unintended consequence of doing what was easiest – avoiding having to find a way to store the global context of each `use`d file, instead recomputing it at each module instantiation and function call.

Fixing the issue is easy, but I've not attached a patch as there is a remote chance that the behavior is intended, and because there is also a remote chance of this breaking / changing the behavior of some existing code that could be relying on this.

Here is the most likely (although very contrived) theoretical scenario I've come up with that would change behavior by fixing this:

**_lib.scad**_

```
var = $fn * 5;
module m() circle($fn=var);
```

And used:

```
use<lib.scad>
module n() {
    $fn = 7;
    m(); // generates a circle with 35 sides
}
n(); 
```

The fix would make m() generate a circle with 25 sides (default $fn*5).

That behavior is a bit inconsistent, as changing `use` -> `include` already today makes the circle get 25 sides.

Is this behavior intended, and are there any more reasonable use-cases? 
##

**Comment by adrianVmariano** [comment]
I stumbled across this recently and was surprised to learn about this old bug report.  

The run time implications of this bug are quite serious:  in a recent test I improved my code run time from 10.5 minutes to 5 seconds by commenting out high level examples in a library file.

**Comment by nophead** [comment]
Yes I moved some test code from a library module into separate test file that used the library and got a speed up.

**Comment by rcolyer** [comment]
I prepared a testcase to help benchmark and illustrate the behavior of this recomputing issue:

slow.scad:
```
//include <slowused.scad>
use <slowused.scad>

x = [for (i=[0:100000]) let($i=i) foo(i)];
echo(x[0]);
```

slowused.scad:
```
howslow = 25;
demodata1 = [for (i=[1:howslow]) rands(0, i, 1)[0]];
$demodata2 = [for (i=[1:howslow]) rands(0, i, 1)[0]];
$i = is_undef($i) ? -1 : $i;
function output2(s) = echo(s) 0;
function output(s) =
  $i == -1 ? output2(str("Initial pass ", s)) :
  $i < 5 ? output2(str("Reprocessed ", s, " time ", $i)) :
  $i == 5 ? output2("Suppressing further output") : 0;
output1 = output("normal variables");
$output2 = output("special variables");

function foo(i) = i+1;
```

This testcase was carefully constructed with the logic needed to misbehave identically in the master branch and in recent releases, separating this remaining performance behavior from the recent changes, in case comparisons across versions are desired to resolve this.  Commenting the use and uncommenting the include produces the correct one-pass behavior:

For include:
```
ECHO: "Initial pass normal variables"
ECHO: "Initial pass special variables"
ECHO: 1
...
Total rendering time: 0:00:00.146
```

However, currently for the use case:
```
ECHO: "Reprocessed normal variables time 0"
ECHO: "Reprocessed special variables time 0"
ECHO: "Reprocessed normal variables time 1"
ECHO: "Reprocessed special variables time 1"
ECHO: "Reprocessed normal variables time 2"
ECHO: "Reprocessed special variables time 2"
ECHO: "Reprocessed normal variables time 3"
ECHO: "Reprocessed special variables time 3"
ECHO: "Reprocessed normal variables time 4"
ECHO: "Reprocessed special variables time 4"
ECHO: "Suppressing further output"
ECHO: "Suppressing further output"
ECHO: 1
...
Total rendering time: 0:00:06.532
```

Increase the value "howslow = 25;" in the used file if you want to see how slow the function calls to the "add one" function can get from the top-level reprocessing in the use case.  I started my testing with howslow = 10000; but this bug has such an impact that one has to turn it down quite a bit to get a tolerably usable testcase.

**Comment by coryrc** [comment]
Given `use<>` is documented to do exactly this and #6131 introduces a fast alternative, I think this should be closed as WAI.

**Comment by adrianVmariano** [comment]
Where is use documented to "do exactly this?"  And even if it is documented that way, that documentation may simply be descriptive to warn users of this behavior.  What SHOULD it do that?   As noted above, "This is not intended - top-level expressions in use'd libraries should only be evaluated once."  

Or are you saying that use should be deprecated and replaced with name spaces and that makes this issue irrelevant?

**Comment by jordanbrown0** [comment]
>Given use<> is documented to do exactly this

My first thought was "It is?".  My second thought was "What doofuss documented that?".  My third thought was "looks like it was me".  Sigh.

@adrianVmariano https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/Include_Statement#use_%3Cfilename%3E

But that was only in March of this year.  I don't remember my exact thought process, except that I was probably describing what was true rather than any intent.  (Shame on me; I've told people numerous times not to do that.)

#6131 might well render this irrelevant, but that's only when and if it merges; until then, this should stay open.

**Comment by jordanbrown0** [comment]
I added a caution to that entry in the documentation, to somewhat stem any future dependence on it until we determine its ultimate fate.

**Comment by coryrc** [comment]
> Where is use documented to "do exactly this?" And even if it is documented that way, that documentation may simply be descriptive to warn users of this behavior. What SHOULD it do that? As noted above, "This is not intended - top-level expressions in use'd libraries should only be evaluated once."

That was 11 years ago. At this point it breaks things to change (i.e. #3881) and... why? why change? Nobody who wants performance is using `use<>` so no matter what code needs to change, why break however many people use `use<>` as it works now? The implementation is small and well-separated, so it's not adding a lot of tech debt to keep it, and the semantics are such that it won't preclude optimizing the rest of the implementation (compare to Python semantics and the CPython implementation, which prevent a lot of optimization other languages/implementations use).

> Or are you saying that use should be deprecated and replaced with name spaces and that makes this issue irrelevant?

Well, I don't know if deprecated; at least discouraged. Otherwise, yes.

**Comment by coryrc** [comment]
btw the lookup in #6131 for `builtins::circle()` is way faster than bosl2's trick of `use<>` to get the builtin circle; just avoiding the cache hit for searching a linked list of 5 lexical contexts on the heap is going to be a good performance win even before adding any optimization.

**Comment by coryrc** [comment]
> [#6131](https://github.com/openscad/openscad/pull/6131) might well render this irrelevant, but that's only when and if it merges; until then, this should stay open.

It should only stay open if you plan to break some users of `use<>` who have been using it this way for at least 11 years. Similar reason why asterisk is still used for multiplication on computer calculators and the Control key is labeled Ctrl and https://linuxreviews.org/WE_DO_NOT_BREAK_USERSPACE and Ctrl-X is cut: don't break your users unless you've got a good reason.

**Comment by MichaelAtOz** [comment]
> I think this should be closed as WAI.

+1, after rereading #3881 and remembering how all the foo()bar() broke my brain I won't be re-arguing anything beyond +1.

p.s. Rereading I note that someone needs to consider whether `-D var=val` command line parameter, which adds a line to main-file EOF, and is the genesis of the 'last assignment wins' style, is not broken, with all the scoping discussions going on.  
I tried to follow the permutations, but my brain is still broken by my medications, and isn't following all the `fn(f)`'s.
Noting some mis-use it as `-D <any OpenSCAD syntax>`. i.e. could `-D namespace <something>` or `-D using <something>` work.
(such discussion not to be posted here)

**Comment by t-paul** [comment]
The very first reply says "This is not intended - top-level expressions in use'd libraries should only be evaluated once."

It was a bug and still is one, so if anyone knows how to fix it, it should be fixed finally. We are not the Linux kernel.

Also did you see the "unless you've got a good reason"? BOSL says "we can't even use the feature as broken as it is", that's not counting as argument?

**Comment by jordanbrown0** [comment]
Right now, we've broken some applications by not including top-level $ variables in use<> files in the dynamic scope for functions and modules in those files.

We've got the worst of all worlds:  broken applications, and no performance fix.

We should either revert the behavior, unbreak the applications, and admit that use<> is always going to be sort of crippled, or accept the compatibility hit, break the applications, and fix the performance.

**Comment by nophead** [comment]
I agree, but it is possible to revert the $variable change and then fix the performance issue by only re-evaluating expressions with $variables in them. Not trivial because they have to be tracked in indirect uses, but not impossible. It would require a data structure for each expression that stores is last value, and a flag to say if it depends on  $variables as a minimum. A more sophisticated solution would have a list of the $variables it depends on and their last value.

**Comment by coryrc** [comment]
> The very first reply says "This is not intended - top-level expressions in use'd libraries should only be evaluated once."
> 
> It was a bug and still is one, so if anyone knows how to fix it, it should be fixed finally. We are not the Linux kernel.

There is no language spec, it is what it does. And what it does is recompute all assignments on every boundary cross and has done so for over 11 years.

> Also did you see the "unless you've got a good reason"? BOSL says "we can't even use the feature as broken as it is", that's not counting as argument?

No, because BOSL2 can use `namespace` which has better semantics and faster implementation. How BOSL2 uses `use<>` today is not affected by this issue because it has no top-level assignments and nothing proposed will make its use of `use<>` any faster or slower than today.

So, no matter what, BOSL2's code has to change to either take advantage of either `use<>` becoming fast or `namespace` being released and it's the same amount of work to use either. So there's zero to gain by breaking `use<>`.

**Comment by t-paul** [comment]
> No, because BOSL2 can use `namespace` which has better semantics and faster implementation. How BOSL2 uses `use<>` today is not affected by this issue because it has no top-level assignments and nothing proposed will make its use of `use<>` any faster or slower than today.

So we can't fix a long standing bug because it's old and user level code must never change, but we declare the solution is to use a different feature that needs the user level code to change so that use<> can be deprecated at some point?

That makes no sense whatsoever.

**Comment by coryrc** [comment]
> So we can't fix a long standing bug 

It's not a bug, it's a feature! Nophead's library is dependent on its current semantics.

> because it's old and user level code must never change, but we declare the solution is to use a different feature that needs the user level code to change so that use<> can be deprecated at some point?

No, some people like its semantics. It never needs to be deprecated.

**Comment by coryrc** [comment]
> > No, because BOSL2 can use `namespace` which has better semantics and faster implementation. How BOSL2 uses `use<>` today is not affected by this issue because it has no top-level assignments and nothing proposed will make its use of `use<>` any faster or slower than today.
> 
> So we can't fix a long standing bug

I also want to clarify the above statement because I confused myself. BOSL2 has one `use<>` statement and it's used as a way to access the builtin definitions of the functions and methods it overrides. This file does not include any variable assignments, so it is not affected by this Issue (#782).

> >  nothing proposed will make __its__ use of `use<>` any faster or slower than today.

But as Chris correctly pointed out that there is a *partial* "fix" that is nearly semantically compatible (the only caveat I've thought of, so far, is side effects from `affect()` or `echo()` in assignments may not be rerun) that can improve the performance for other users of `use<>` besides BOSL2, but it comes with a large complexity increase.

**Comment by t-paul** [comment]
> No, some people like its semantics. It never needs to be deprecated.

"Some people" are not doing the work of maintaining OpenSCAD. It's simply not viable to continuously maintain multiple combinations of previous bugs. OpenSCAD is traditionally very very conservative with changes. There has to be a line *somewhere*. Reintroducing bad bugs and then either not fixing this one here, or making the bug even more convoluted and complicated is not a sensible option, even if we call it "partial".

**Comment by coryrc** [comment]
> > No, some people like its semantics. It never needs to be deprecated.
> 
> Some people are not doing the work of maintaining OpenSCAD. It's simply not viable to continuously maintain multiple all combination of previous bugs.

I am now intimately familiar with how names and scoping work because of #6131. The implementation of `use<>`, as it is, is mostly quite a small chunk (mostly lines [79-132 of ScopeContext.cc](https://github.com/openscad/openscad/blob/2382389c7f09ba6789ba027b17ce442dfd85c660/src/core/ScopeContext.cc#L79)). (I even simplified it in #6131 to be shorter!). There's that, a parsing rule and the AST node for use<>. The rest it reuses: SourceFile, SourceFileCache, FileContext are all needed for other purposes. EDIT: also `registerUse()` method of SourceFile.

Adding what Chris suggests would be a huge maintenance burden as I pointed out, but leaving it alone as it functions now is **not**.

> OpenSCAD is traditionally very very conservative with changes.

It's not, because someone broke `use<>`'s dynamic scope in a commit called "refactoring" and so far you have said you will not allow it to be put back. I will make the small change needed to restore the dynamic scope inheritance if you will merge it.

>There has to be a line _somewhere_. Reintroducing bad bugs and then either not fixing this one here, or making the bug even more convoluted and complicated is not a sensible option, even if we call it "partial".

It's not a bug! You can't release something that works one way for 11 years without documentation then say "oops that isn't what I meant it to do" and wildly change it and still call yourself "conservative with changes". If it was fixed in 2014 *when this Issue was filed*, fine, but it's 2025. `use<>` has existed since [at least 2013](https://github.com/openscad/openscad/blame/9a297ecee57549a4eb3919bde4f7b41a548724de/src/modcontext.cc#L28) and even before the [first commit you ever made](https://github.com/openscad/openscad/commit/bc4ace7e2641c67fa0677587d4c34a3d2bb239b6) to OpenSCAD.

There have been 5 stable releases with this functionality! All stable releases in the last **decade** work this way. There's nothing conservative about changing it now.

**Comment by t-paul** [comment]
> It's not, because someone broke `use<>`'s dynamic scope in a commit called "refactoring" and so far you have said you will not allow it to be put back. I will make the small change needed to restore the dynamic scope inheritance if you will merge it.

The commit may have a wrong title, we have lots of those. It does not reflect what happened at all. I will stop now wasting my time running in circles again as it's obvious that nobody cares what I have to say.

**Comment by adrianVmariano** [comment]
I don't claim to understand all the scoping subtleties, so maybe I'm missing something here, but my understanding is that a change was made affecting use and $ variables.  This change breaks nophead's library and is already in the nightly build.   Meanwhile, the bug in question, where used ordinary variables are re-evaluated has NOT been fixed.  So I don't understand the link between these things.  It's difficult to conceive of a reason why it's desirable to re-evaluate top level assignments in a language where variables are immutable.  

The point about BOSL2 is not that it relies on use but that it avoids use.  We abandoned use because it was unusable do to the performance issues created by the top level re-assignments.  It's hard to see how this is anything other than a bug.  Doesn't matter if anybody somehow managed to make their code depend on it.  Or if it's been around for 11 years---wrong code doesn't become right just because it's old.   (The argument about whether it's a bug that needs to be conserved is a separate matter.)

**Comment by coryrc** [comment]
> I don't claim to understand all the scoping subtleties, so maybe I'm missing something here, but my understanding is that a change was made affecting use and $ variables. This change breaks nophead's library and is already in the nightly build.

Correct, the subtlety is [does the function called in a `use<>`d file take the dynamic scope of the caller or of the top-level of the `use<>` file](https://github.com/openscad/openscad/issues/3881#issuecomment-3246905549). Partially related to this bug, because if you don't re-evaluate the `use<>`d variables, then there's no dynamic scope to inherit from anyway.

> Meanwhile, the bug in question, where used ordinary variables are re-evaluated has NOT been fixed. So I don't understand the link between these things. It's difficult to conceive of a reason why it's desirable to re-evaluate top level assignments in a language where variables are immutable.

You're forgetting dynamic scope. If your `use<>`d file is just `a=$a; function getA() = a;`, then you get different answers to:

```
$a = 1;
use<getA.scad>
echo(getA());
```

When do you evaluate `use<>`? Does it inherit the top-level's dynamic scope? What if you `use<>` before and after `$a=1`?

> The point about BOSL2 is not that it relies on use but that it avoids use. We abandoned use because it was unusable do to the performance issues created by the top level re-assignments.

Right, so you need to change the codebase to make use of something without performance issues. Does it matter if you call it `namespace` or `use`? All you want is something fast and isolating. #6131 does that.

> It's hard to see how this is anything other than a bug.

Do you still feel that way after seeing the above?

> Doesn't matter if anybody somehow managed to make their code depend on it. Or if it's been around for 11 years---wrong code doesn't become right just because it's old. (The argument about whether it's a bug that needs to be conserved is a separate matter.)

There's no such thing as an objective "wrong code". When used with 2021.01 (or even older versions!) it works as nophead wants it to. That sounds a lot closer to an objective "right" than people saying 11 years after-the-fact "that's not what I meant".

Microsoft became one of top 10 most valuable companies in the world because they prioritized keeping customer "[wrong code](https://news.ycombinator.com/item?id=7809433)" working.
